src/storage/v3/shard_rsm.cpp

```diff
@@ -934,12 +935,115 @@ msgs::ReadResponses ShardRsm::HandleRead(msgs::ScanVerticesRequest &&req) {
+      std::map<PropertyId, memgraph::msgs::Value> ret;
+      auto property_results = edge.Properties(View::OLD);
+      if (property_results.HasError()) {
+        spdlog::debug("Encountered an error while trying to get out-going EdgeAccessors. Transaction id: {}",
+                      req.transaction_id.logical_id);
+        return LocalError{};
+      }
+
+      auto &properties = property_results.GetValue();
+      std::transform(properties.begin(), properties.end(), std::inserter(ret, ret.end()), [](auto &prop) {
+        return (std::make_pair(prop.first, FromPropertyValueToValue(std::move(prop.second))));
+      });
+
+      return ret;
+    };
+  } else {
+    // TODO(gvolfing) - do we want to set the action_successful here?
+    get_edge_properties = [&req](const memgraph::storage::v3::EdgeAccessor &edge) {
+      std::vector<memgraph::msgs::Value> ret;
+      ret.reserve(req.edge_properties.value().size());
+
+      // TODO(gvolfing) maybe check for the absence of certain properties
+      const auto &edge_properties = req.edge_properties.value();
+      std::transform(edge_properties.begin(), edge_properties.end(), std::back_inserter(ret),
+                     [&edge](const auto &edge_prop) {
+                       return FromPropertyValueToValue(edge.GetProperty(edge_prop, View::OLD).GetValue());
+                     });
+
+      return ret;
+    };
+  }
+
+  // Functions to select connecting edges based on uniquness
+  using EdgeAccessors = std::vector<memgraph::storage::v3::EdgeAccessor>;
+  std::function<EdgeAccessors(EdgeAccessors &&, memgraph::msgs::EdgeDirection)> maybe_filter_based_on_edge_uniquness;
+  if (req.only_unique_neighbor_rows) {
+    maybe_filter_based_on_edge_uniquness = [](EdgeAccessors &&edges,
+                                              memgraph::msgs::EdgeDirection edge_direction) -> EdgeAccessors {
+      std::function<bool(std::unordered_set<const storage::v3::VertexId *> &,
+                         const memgraph::storage::v3::EdgeAccessor &)>
+          check_and_insert_unique_other_vertex;
+      switch (edge_direction) {
+        case memgraph::msgs::EdgeDirection::OUT: {
+          check_and_insert_unique_other_vertex = [](std::unordered_set<const storage::v3::VertexId *> &other_vertex_set,
+                                                    const memgraph::storage::v3::EdgeAccessor &edge_acc) {
+            auto [it, insertion_happened] = other_vertex_set.insert(&edge_acc.ToVertex());
+            return insertion_happened;
+          };
+          break;
+        }
+        case memgraph::msgs::EdgeDirection::IN: {
+          check_and_insert_unique_other_vertex = [](std::unordered_set<const storage::v3::VertexId *> &other_vertex_set,
+                                                    const memgraph::storage::v3::EdgeAccessor &edge_acc) {
+            auto [it, insertion_happened] = other_vertex_set.insert(&edge_acc.FromVertex());
+            return insertion_happened;
```

Comment:
One more thing: the set is based on pointer without a custom comparator, that means it will only differentiate between the vertex ids based on there address, which is different in each edge, even for the same vertices. So I think either a custom comparator or using value semantics is necessary.

Maybe also think about eliminating these functions (`std::function` is much more expensive than a lambda) by getting the vertex id conditionally:
```

const auto vertex_selector = [edge_direction](cosnt EdgeAccessor& edge_acc) {
  switch (edge_direction): {
    case msgs::EdgeDirection::IN:
      return edge_acc.FromVertex();
    case msgs::EdgeDirection::OUT:
      return edge_acc.ToVertex();
    case msgs::EdgeDirection::BOTH:
      LOG_FATAL("This is should never happen, memgraph::msgs::EdgeDirection::BOTH should not be passed here.");
  }
}
```

And then you can use it to get the correct vertex id. In my opinion this solution would have two advantages:
1. It is not using a `std::function` (`check_and_insert_unique_other_vertex`), it will be faster
2. It makes the logic a bit simpler in my opinion.

Reply:
This part of the code was removed with `maybe_filter_based_on_edge_uniquness`.